clickhouse_driver/client.py

```diff
@@ -539,10 +539,19 @@ def insert_dataframe(
                 # raise if any columns are missing from the dataframe
                 diff = set(columns) - set(dataframe.columns)
                 if len(diff):
-                    msg = "DataFrame missing required columns: {}"
-                    raise ValueError(msg.format(list(diff)))
+                    raise ValueError(
+                        f"DataFrame has missing required columns: {list(diff)}"
+                    )
+
+                # TODO: to find a more pandas-idiomatic way
+                data = []
+                for column in columns:
+                    column_values = dataframe[column].values
+                    for idx, col_vals in enumerate(column_values):
+                        if isinstance(col_vals, dict):
+                            column_values[idx] = tuple(col_vals.values())
```

Comment:
What about speed and memory consumption here? Did you test the new code with large dataframes?